### ISSUE TITLE
NC XWIKI Instance config shows "the Nextcloud app is missing"

### ISSUE THREAD

**Issue by sebastiaanveld** [nextcloud/xwiki#2]
Problem:
After setting up XWIKI and Nextcloud according to the available documentation I keep getting an message in the Nextcloud XWIKI Administration section  "the Nextcloud app is missing".

Steps to duplicate: 
* I've setup NC (25.0.5) and XWIKI (tested both 15.1 and 15.2):
* added the XWIKI App to NC
* Added the Nextcloud Application to XWIKI
* Ajusted xwiki.cfg adn added xwiki.authentication.authclass=org.xwiki.contrib.oidc.provider.OIDCBridgeAuth
* Run in NC the XWIKI registration process, which seems to work ok as:
* Added my XWIKI URL, then I'm redifrected to the WIKI and I can copy my generated client key.
* Pasted the key back in the Client ID section in NC.
*  Then when I run a check I see "✓ (XWiki v15.2) (the Nextcloud app is missing)"

![image](https://user-images.githubusercontent.com/37268999/229295707-811a1888-73f7-4bc0-a0cc-1d8942b75d96.png)

**Comment by raphj** [comment]
@sebastiaanveld have you restarted XWiki after this step: "added the XWIKI App to NC"?

**Comment by sebastiaanveld** [comment]
Yes, XWIKI was restarted after adding the Nextcloud App and adding the additional line to xwiki.cfg.
xwiki.authentication.authclass=org.xwiki.contrib.oidc.provider.OIDCBridgeAuth

From some testing I found when I go to https://wiki.virtfields.net/xwiki/oidc/ this shows all "listening urls" (is that the right wording for that?) all shown as unsecure http over port 80 instead of https:<url>:443, could this be the issue? I would expect the wiki expose those as secure links. The wiki is only reachable over https.
	
authorization_endpoint	"http://wiki.virtfields.net:80/xwiki/oidc/authorization"
token_endpoint	"http://wiki.virtfields.net:80/xwiki/oidc/token"
issuer	"http://wiki.virtfields.net:80"
jwks_uri	"http://wiki.virtfields.net:80/xwiki/oidc/jwk"
request_uri_parameter_supported	true
subject_types_supported	
0	"public"
userinfo_endpoint	"http://wiki.virtfields.net:80/xwiki/oidc/userinfo"

**Comment by raphj** [comment]
This http on port 80 is of course an issue. Is XWiki aware that it is accessed from HTTPS?

does your reverse proxy provide the right HTTP headers to XWiki?

My test / local nginx conf includes this for instance:

```
  location / {
    proxy_pass http://localhost:8080/;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Host $host;
  }
```

See also https://forum.xwiki.org/t/how-to-setup-xwiki-https/257

**Comment by sebastiaanveld** [comment]
I'm using HAproxy with forwardfor so that should be fine.

Added to xwiki.cfg:
xwiki.home=https://xwiki.virtfields.net:443/
xwiki.url.protocol=https

Now the endpoints show https over port 443 as expected, but in Nextcloud I still the same error running the check.


```
authorization_endpoint | "https://wiki.virtfields.net:443/xwiki/oidc/authorization"
token_endpoint | "https://wiki.virtfields.net:443/xwiki/oidc/token"
issuer | "https://wiki.virtfields.net:443/"
jwks_uri | "https://wiki.virtfields.net:443/xwiki/oidc/jwk"
request_uri_parameter_supported | true
subject_types_supported |  
0 | "public"
userinfo_endpoint | "https://wiki.virtfields.net:443/xwiki/oidc/userinfo"

</div></div></div></div><!--EndFragment-->
</body>
</html>authorization_endpoint	"[https://wiki.virtfields.net:443/xwiki/oidc/authorization](https://wiki.virtfields.net/xwiki/oidc/authorization)"
token_endpoint	"[https://wiki.virtfields.net:443/xwiki/oidc/token](https://wiki.virtfields.net/xwiki/oidc/token)"
issuer	"[https://wiki.virtfields.net:443/](https://wiki.virtfields.net/)"
jwks_uri	"[https://wiki.virtfields.net:443/xwiki/oidc/jwk](https://wiki.virtfields.net/xwiki/oidc/jwk)"
request_uri_parameter_supported	true
subject_types_supported	
0	"public"
userinfo_endpoint	"[https://wiki.virtfields.net:443/xwiki/oidc/userinfo](https://wiki.virtfields.net/xwiki/oidc/userinfo)"
```

**Comment by raphj** [comment]
I think I found the issue.

The Nextcloud application checks if the Nextcloud.WebHome space exists. It does this unauthenticated.
You need to allow a guest user to access `Nextcloud.WebHome`.

`https://wiki.virtfields.net:443/xwiki/rest/wikis/xwiki/spaces/Nextcloud/pages/WebHome` needs to return a HTTP 200 status.

I get:

```
$ curl https://wiki.virtfields.net:443/xwiki/rest/wikis/xwiki/spaces/Nextcloud/pages/WebHome                                                                                                                                                                                                                                                                               
{"code":401,"contactEmail":null,"description":"The request requires user authentication","homeRef":"/","reasonPhrase":"Unauthorized","uri":"http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.2"}

 $ curl -I https://wiki.virtfields.net:443/xwiki/rest/wikis/xwiki/spaces/Nextcloud/pages/WebHome                                                                                                                                                                                                                                                                          [130]
HTTP/1.1 401 
content-script-type: text/javascript
set-cookie: JSESSIONID=B37E8FE2AB59A188D22FF4819DDF9465; Path=/xwiki; HttpOnly
date: Sun, 02 Apr 2023 21:51:14 GMT
accept-ranges: bytes
server: Restlet-Framework/2.3.12
xwiki-version: 15.2
xwiki-form-token: KZxDUFUA8vLyxHj9Aumc0w
content-type: text/html;charset=utf-8
content-language: en
transfer-encoding: chunked
```

You might be able to set this in the user & rights preference of the `Nextcloud` space. Also mind the Administration option to prevent Guest users from accessing your wiki no matter what (" Prevent unregistered users from viewing pages, regardless of the page rights"). This option is currently not supported by the Nextcloud integration. 

![image](https://user-images.githubusercontent.com/3817365/229381117-69192514-2146-42a7-beb5-1047ae5ecbff.png)

You can prevent Guest users from accessing your wiki by setting the necessary rights to the spaces and pages you want to protect thanks to the "Rights: Page & Children" section in the Page Administration of these spaces.

**Comment by raphj** [comment]
>Cool, that did the trick!

Nice! :-)

This will 

> I would expect the user to list in the users/groups section so the XWIKI admin can assign a page structure to the user

Nope, it does not work like this. Each Nextcloud user will be able to login to XWiki from the "XWiki section" of their personal setting of Nextcloud. They will then be able to see their own content as their own XWiki user.

So if you want to assign page structure to specific user you would just set such rights on the relevant spaces / pages for the native XWiki user accounts of these users. Hope that makes sense.

> If I search for content from NC I get this:
> ![image](https://user-images.githubusercontent.com/37268999/229444447-cacc400e-c50a-41d5-b479-6854aabee6e0.png)

This is expected. You landed on the XWiki.WebHome space, NC is displaying you the default content of this space. You can see this content on XWiki by going to http://wiki.virtfields.net/xwiki/bin/view/XWiki/. `XWiki` is a technical space not really intended to be visited. It stores administrative settings, users and groups, some provided macros and features, among other things.

**Comment by raphj** [comment]
It works a bit like "Sign in with GitHub / Microsoft / Google" on 3rd party websites

**Comment by sebastiaanveld** [comment]
```
    how does that account mapping take place

With OpenID-Connect. Users are logged on XWiki. Nextcloud will ask for permission to the user to access to XWiki on their behalf.
```

Ok, that makes sense.

Now I'm logged in as a non admin in nextcloud, so I'm a user:

- I click the XWIKI icon on the top bar
- Click the Get Access button
- Now the message appears again the Nextcloud App is not installed in XWIKI

![image](https://user-images.githubusercontent.com/37268999/229504428-34eda3ef-a5c2-4db2-9ca3-dc7ca1dbae29.png)

for this user I also created a local XWIKI account. Even if I authenticate the user upfront in the wiki I still get the same error.

**Comment by raphj** [comment]
Is this still true?

> Cool, that did the trick! The admin page now shows the App is installed and I can register the wiki for the user.

Maybe remove and re-pair xwiki with Nextcloud again and see if you still have the message?

**Comment by sebastiaanveld** [comment]
Yes: from what I see it works when I'm logged into XWIKI as an admin.

But when I'm logged into XWIKI as an user or not logged into XWIKI at all NC message shows the logged in user should be an admin

![image](https://user-images.githubusercontent.com/37268999/229553698-6b8c2b8c-b66e-4565-a54e-729aa8007c9c.png)

**Comment by raphj** [comment]
Right, that's a bug. I'll look into it. A non-admin nextcloud user can't login.

https://github.com/nextcloud/xwiki/issues/6

**Comment by sebastiaanveld** [comment]
Thx, we'll follow up futher on #6 where I added some more context.
